selene-lib/src/rules/mismatched_arg_count.rs

```diff
@@ -0,0 +1,425 @@
+};
+use id_arena::Id;
+
+pub struct MismatchedArgCountLint;
+
+impl Rule for MismatchedArgCountLint {
+    type Config = ();
+    type Error = Infallible;
+
+    fn new(_: Self::Config) -> Result<Self, Self::Error> {
+        Ok(MismatchedArgCountLint)
+    }
+
+    fn pass(&self, ast: &Ast, _context: &Context) -> Vec<Diagnostic> {
+        let scope_manager = ScopeManager::new(ast);
+
+        // Firstly visit the AST so we can map the variables to their required parameter counts
+        let mut definitions = HashMap::new();
+        let mut definitions_visitor = MapFunctionDefinitionVisitor {
+            scope_manager: &scope_manager,
+            definitions: &mut definitions,
+            blacklisted_variables: HashSet::new(),
+        };
+        definitions_visitor.visit_ast(&ast);
+
+        let mut visitor = MismatchedArgCountVisitor {
+            mismatched_arg_counts: Vec::new(),
+            scope_manager,
+            definitions,
+        };
+
+        visitor.visit_ast(&ast);
+
+        visitor
+            .mismatched_arg_counts
+            .iter()
+            .map(|mismatched_arg| {
+                Diagnostic::new_complete(
+                    "mismatched_arg_count",
+                    format!(
+                        "this function takes {}",
+                        mismatched_arg
+                            .parameter_count
+                            .to_message(mismatched_arg.num_provided)
+                    ),
+                    Label::new_with_message(
+                        mismatched_arg.call_range,
+                        mismatched_arg.parameter_count.to_string(),
+                    ),
+                    Vec::new(),
+                    vec![Label::new_with_message(
+                        mismatched_arg.function_definition_range,
+                        "note: function defined here".to_owned(),
+                    )],
+                )
+            })
+            .collect()
+    }
+
+    fn severity(&self) -> Severity {
+        Severity::Error
+    }
+
+    fn rule_type(&self) -> RuleType {
+        RuleType::Correctness
+    }
+}
+
+struct MismatchedArgCount {
+    parameter_count: ParameterCount,
+    num_provided: PassedArgumentCount,
+    call_range: (usize, usize),
+    function_definition_range: (usize, usize),
+}
+
+#[derive(Clone, Copy, Debug)]
+enum ParameterCount {
+    /// A fixed number of parameters are required: `function(a, b, c)`
+    Fixed(usize),
+    /// Some amount of fixed parameters are required, and the rest are variable: `function(a, b, ...)`
+    Minimum(usize),
+    /// A variable number of parameters can be provided: `function(...)`
+    Variable,
+}
+
+impl ParameterCount {
+    /// Calculates the number of required parameters that must be passed to a function
+    pub fn from_function_body(function_body: &ast::FunctionBody) -> Self {
+        let mut necessary_params = 0;
+
+        for parameter in function_body.parameters() {
+            match parameter {
```

Comment:
If it doesn't warn before we implement the full-moon update that makes this exhaustive, give this a `_` branch.

Reply:
This lints with `unreachable_pattern`